// @api-platform/api-doc-parser/lib/cjs/swagger/handleJson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeTrailingSlash = void 0;
var tslib_1 = require("tslib");
var lodash_get_1 = tslib_1.__importDefault(require("lodash.get"));
var inflection_1 = require("inflection");
var Field_1 = require("../Field");
var Resource_1 = require("../Resource");
var getResources_1 = tslib_1.__importDefault(require("../utils/getResources"));
var getType_1 = tslib_1.__importDefault(require("../openapi3/getType"));
var removeTrailingSlash = function (url) {
    if (url.endsWith("/")) {
        url = url.slice(0, -1);
    }
    return url;
};
exports.removeTrailingSlash = removeTrailingSlash;
function default_1(response, entrypointUrl) {
    var paths = (0, getResources_1.default)(response.paths);
    return paths.map(function (path) {
        var splittedPath = (0, exports.removeTrailingSlash)(path).split("/");
        var name = (0, inflection_1.pluralize)(splittedPath[splittedPath.length - 2]);
        var url = "".concat((0, exports.removeTrailingSlash)(entrypointUrl), "/").concat(name);
        var title = (0, inflection_1.classify)(splittedPath[splittedPath.length - 2]);
        if (!response.definitions) {
            throw new Error(); // @TODO
        }
        var definition = response.definitions[title];
        if (!definition) {
            throw new Error(); // @TODO
        }
        var description = definition.description;
        var properties = definition.properties;
        if (!properties) {
            throw new Error(); // @TODO
        }
        var fieldNames = Object.keys(properties);
        var requiredFields = (0, lodash_get_1.default)(response, ["definitions", title, "required"], []);
        var fields = fieldNames.map(function (fieldName) {
            var property = properties[fieldName];
            return new Field_1.Field(fieldName, {
                id: null,
                range: null,
                type: (0, getType_1.default)((0, lodash_get_1.default)(property, "type", ""), (0, lodash_get_1.default)(property, "format", "")),
                reference: null,
                embedded: null,
                required: !!requiredFields.find(function (value) { return value === fieldName; }),
                description: property.description || "",
            });
        });
        return new Resource_1.Resource(name, url, {
            id: null,
            title: title,
            description: description,
            fields: fields,
            readableFields: fields,
            writableFields: fields,
        });
    });
}
exports.default = default_1;
